src/Features/Core/Portable/AddParameter/AddParameterService.cs

```diff
@@ -0,0 +1,344 @@
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.AddParameter
+{
+    internal class AddParameterService
+    {
+        private AddParameterService() { }
+
+        public static AddParameterService Instance = new AddParameterService();
+
+        /// <summary>
+        /// Checks if there are indications that there might be more than one declarations that need to be fixed.
+        /// The check does not look-up if there are other declarations (this is done later in the CodeAction).
+        /// </summary>
+        public bool HasCascadingDeclarations(IMethodSymbol method)
+        {
+            // Don't cascade constructors
+            if (method.IsConstructor())
+            {
+                return false;
+            }
+
+            // Virtual methods of all kinds might have overrides somewhere else that need to be fixed.
+            if (method.IsVirtual || method.IsOverride || method.IsAbstract)
+            {
+                return true;
+            }
+
+            // If interfaces are involved we will fix those too
+            // Explicit interface implementations are easy to detect
+            if (method.ExplicitInterfaceImplementations.Length > 0)
+            {
+                return true;
+            }
+
+            // For implicit interface implementations lets check if the characteristic of the method
+            // allows it to implicit implement an interface member.
+            if (method.DeclaredAccessibility != Accessibility.Public)
+            {
+                return false;
+            }
+
+            if (method.IsStatic)
+            {
+                return false;
+            }
+
+            // Now check if the method does implement an interface member
+            if (method.ExplicitOrImplicitInterfaceImplementations().Length > 0)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Adds a parameter to a method.
+        /// </summary>
+        /// <param name="invocationDocument"></param>
+        /// <param name="method"></param>
+        /// <param name="newParamaterType"></param>
+        /// <param name="refKind"></param>
+        /// <param name="parameterName"></param>
+        /// <param name="newParameterIndex">null to add as the final parameter</param>
+        /// <param name="fixAllReferences"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<Solution> AddParameterAsync(
+            Document invocationDocument,
+            IMethodSymbol method,
+            ITypeSymbol newParamaterType,
+            RefKind refKind,
+            string parameterName,
+            int? newParameterIndex,
+            bool fixAllReferences,
+            CancellationToken cancellationToken)
```

Comment:
can you assert that you haven't changed these methods from their originating location? If you have changed them, can you sum up the types of changes you made?

Reply:
(this applies to all the methods moved)

Reply:
Only two methods were changed:
FixAsync was split into two methods, FixAsync (which remained in AbstractAddParameterCodeFixProvider) and AddParameterAsync.

FixAsync contains the first two lines of the original FixAsync, plus also the logic to calculate at what index to add the new parameter, which was orignially done at a later stage:

```csharp
// currently done here:
            var newParameterIndex = isNamedArgument ? (int?)null : argumentList.IndexOf(argument);

//originally done here:
                    var insertionIndex = isNamedArgument
                        ? existingParameters.Count
                        : argumentList.IndexOf(argument);
```

The only other change is removing the unused `TArgumentSyntax argument` parameter from  `AddParameter()`.